sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala

```diff
@@ -346,4 +346,23 @@ private[sql] object CatalogV2Util {
       }
     }
   }
+
+  def failNullType(dt: DataType): Unit = {
+    def containsNullType(dt: DataType): Boolean = dt match {
+      case ArrayType(et, _) => containsNullType(et)
+      case MapType(kt, vt, _) => containsNullType(kt) || containsNullType(vt)
+      case StructType(fields) => fields.exists(f => containsNullType(f.dataType))
+      case _ => dt.isInstanceOf[NullType]
+    }
+    if (containsNullType(dt)) {
+      throw new AnalysisException(
+        "Cannot create tables with VOID type.")
```

Comment:
Ur, are we going to expose `VOID` here? Maybe, `NULL` is better at this error message? Technically, this function is recognizing `NullType`, not `Void` type in Apache Spark Catalyst module.

Reply:
@cloud-fan ask me to use VOID from NULL. Personally I agree to use VOID even through NullType is a Type but "null" has a literal syntax. Every time when we see a "null", we need to finger out it's a data type or literal. It's a little ambiguity. So @cloud-fan also suggests me to change the simpleString of `NullType` to "void". So make the symbol "null" only be literal.

Reply:
But I am not sure the behaviour in other database managements. Maybe `NULL` is a data type and `null` is a literal?

Reply:
I will check some database system to see their behaviours, wait a minute.

Reply:
Thank you for checking, @LantaoJin . This is also confusing to me. From Hive side, `VOID` is a type name. But, Apache Spark side, 
- We have `null` literal: https://spark.apache.org/docs/latest/sql-ref-literals.html#null-literal
- We have only `nullable` types: https://spark.apache.org/docs/latest/sql-ref-datatypes.html

Reply:
If we use `VOID` here in Apache Spark exception type, it looks to me like we declare `VOID` is one of Apache Spark type. But, please follow, @cloud-fan and @gatorsmile 's decision.

Reply:
AFAIK some databases use UNKNOWN to represent the null type. Maybe UNKNOWN is better as it's not a data type so we don't need to document it.

Reply:
I only tested in PostgreSQL and MySQL.
No one can run
`CREATE TABLE t (null_col NULL);`
For 
`CREATE TABLE t2 AS SELECT NULL as null_col`
PostgreSQL shows
```
                Table "public.t2"
  Column  | Type | Collation | Nullable | Default
----------+------+-----------+----------+---------
 null_col | text |           |          |
```
And MYSQL shows
```
  Column  |    Type   | Nullable  | Default
----------+-----------+-----------+----------
 null_col | binary(0) |    YES    |  NULL
```

Reply:
So there are two points to be determinated.
1. What is the exception message?
`"Cannot create tables with VOID type."` or `"Cannot create tables with NULL type."` or "Cannot create tables with UNKNOWN type."
2. What is the `simpleString` of class `NullType` ?
`null` (current) or `void`(hive) or `unknown`

Reply:
I'm in favor of UNKNOWN/unknown, as it indicates it's not a real data type. But I'm open to other options. cc @gatorsmile @HyukjinKwon @maropu @viirya @bart-samwel

Reply:
`unknown` looks okay to me. Since `null` is mainly used to represent a literal in spark, I think its better to avoid using it for data types. Also, I think `void` is a word that is rarely used in relational databases.

Reply:
So basically, looks like current change allows to parse `void` as `NullType`, but if it is used in a command like creating table, we throw exception?


One question, if we don't allow void/unknown type, is there good reason to accept in parser?

Because we don't have real `void` data type, to say "unknown" type in error message looks okay.

Reply:
Yes. We throw exception in command like creating table.

> One question, if we don't allow void/unknown type, is there good reason to accept in parser?

VOID is a legacy Hive table type which may be read by Spark. So in #28935, I map `case ("void", Nil) => HiveVoidType` in Parser, but @cloud-fan suggested me to re-use NullType. So the syntax seems we accepted void/unknown in parser. Actually, first, we accept the legacy Hive type VOID in parser and two, we fobid creating table with void/null.

Reply:
Looks like `unknown` is accepted here. I am going to patch it. Any more comments @dongjoon-hyun @cloud-fan @gatorsmile

Reply:
Let's go with `unknown` then